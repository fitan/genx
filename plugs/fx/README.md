# @fx Plugin - Uber FX 依赖注入自动化

基于 `@do` 插件的设计思路，为 Uber FX 框架提供自动化的依赖注入配置。

## 🎯 功能特性

- ✅ 自动扫描 `@fx` 注解的函数
- ✅ 生成 fx.Option 配置
- ✅ 支持 Provide、Invoke、Decorate
- ✅ 支持接口绑定和分组
- ✅ 智能包名冲突处理
- ✅ 确定性排序保证一致性

## 📝 使用方式

### 1. 服务提供者 (Provide)

```go
// services/user_service.go

// @fx provide
func NewUserService(db *gorm.DB, logger *slog.Logger) *UserService {
    return &UserService{db: db, logger: logger}
}

// @fx provide UserServiceInterface  // 绑定到接口
func NewUserServiceImpl(db *gorm.DB) UserServiceInterface {
    return &UserService{db: db}
}

// @fx provide "" handlers  // 加入到 handlers 组
func NewUserHandler(svc *UserService) *UserHandler {
    return &UserHandler{svc: svc}
}
```

### 2. 启动时调用 (Invoke)

```go
// app/startup.go

// @fx invoke
func StartHTTPServer(handler *gin.Engine, lifecycle fx.Lifecycle) {
    srv := &http.Server{
        Addr:    ":8080",
        Handler: handler,
    }
    
    lifecycle.Append(fx.Hook{
        OnStart: func(ctx context.Context) error {
            go srv.ListenAndServe()
            return nil
        },
        OnStop: func(ctx context.Context) error {
            return srv.Shutdown(ctx)
        },
    })
}

// @fx invoke
func InitDatabase(db *gorm.DB) error {
    return db.AutoMigrate(&User{}, &Order{})
}
```

### 3. 装饰器 (Decorate)

```go
// middleware/logging.go

// @fx decorate UserServiceInterface  // 装饰接口
func NewLoggingUserService(svc UserServiceInterface, logger *slog.Logger) UserServiceInterface {
    return &LoggingUserService{
        next:   svc,
        logger: logger,
    }
}

// @fx decorate
func NewCachedUserService(svc *UserService, cache *redis.Client) *UserService {
    return &CachedUserService{
        UserService: svc,
        cache:       cache,
    }
}
```

### 4. 应用入口标记

```go
// cmd/app/main.go

// @fx app
func main() {
    // 这个函数体不重要，只是标记在这个包生成 fx_app.go
    app := NewApp()
    app.Run()
}
```

## 🏗️ 生成的代码

### 自动生成的 `cmd/app/fx_app.go`:

```go
package app

import (
    "go.uber.org/fx"
    services "myproject/services"
    middleware "myproject/middleware"
    app1 "myproject/app"
)

// FxOptions 返回所有自动收集的 fx.Option
func FxOptions() []fx.Option {
    return []fx.Option{
        // Provide 选项
        fx.Provide(services.NewUserService),
        fx.Provide(fx.Annotate(
            services.NewUserServiceImpl,
            fx.As(new(UserServiceInterface)),
        )),
        fx.Provide(fx.Annotate(
            services.NewUserHandler,
            fx.ResultTags(`group:"handlers"`),
        )),
        
        // Decorate 选项
        fx.Decorate(fx.Annotate(
            middleware.NewLoggingUserService,
            fx.As(new(UserServiceInterface)),
        )),
        fx.Decorate(middleware.NewCachedUserService),
        
        // Invoke 选项
        fx.Invoke(app1.StartHTTPServer),
        fx.Invoke(app1.InitDatabase),
    }
}

// NewApp 创建配置好的 fx.App
func NewApp(opts ...fx.Option) *fx.App {
    allOpts := append(FxOptions(), opts...)
    return fx.New(allOpts...)
}
```

## 🚀 使用生成的代码

```go
// cmd/app/main.go
func main() {
    // 方式1: 使用生成的 NewApp
    app := NewApp(
        fx.Provide(NewConfig),  // 额外的配置
        fx.WithLogger(func() fxevent.Logger {
            return &fxevent.ConsoleLogger{W: os.Stdout}
        }),
    )
    
    // 方式2: 手动组合
    app := fx.New(
        append(FxOptions(), 
            fx.Provide(NewConfig),
            fx.WithLogger(...),
        )...,
    )
    
    app.Run()
}
```

## 🔗 与其他插件协作

其他 genx 插件生成的代码可以自动集成：

```go
// @log 插件生成的代码会包含 @fx provide 注解
// @fx provide
func NewLogging(logger *slog.Logger) func(next Service) Service {
    return func(next Service) Service {
        return &logging{logger: logger, next: next}
    }
}

// @temporal 插件生成的代码
// @fx provide
func InitTemporal(client client.Client, worker worker.Worker) (*Temporal, error) {
    // ...
}
```

## 📊 注解语法

### @fx provide [interface] [group]
- `interface`: 可选，绑定到指定接口类型
- `group`: 可选，加入到指定组

### @fx invoke
- 标记需要在应用启动时调用的函数

### @fx decorate [interface]
- `interface`: 可选，装饰指定接口类型

### @fx app
- 标记应用入口，在此包生成 fx_app.go

## 🎯 优势

1. **零样板代码**: 自动生成所有 fx.Option 配置
2. **类型安全**: 编译时检查依赖关系
3. **模块化**: 每个包独立定义依赖
4. **可扩展**: 支持额外的 fx.Option
5. **一致性**: 确定性排序保证构建一致性

## 🔄 迁移指南

### 从 @do 迁移到 @fx

```go
// 原来的 @do 注解
// @do provide
func NewUserService() *UserService { ... }

// 改为 @fx 注解  
// @fx provide
func NewUserService() *UserService { ... }

// 原来的手动初始化
func main() {
    injector := do.New()
    doInit(injector)
    svc := do.MustInvoke[*UserService](injector)
}

// 改为 fx 方式
func main() {
    app := NewApp()
    app.Run()
}
```

这样你就可以享受 fx 的强大功能，同时保持 genx 的自动化便利！
