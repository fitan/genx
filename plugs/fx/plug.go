package fx

import (
	"fmt"
	"log/slog"
	"path/filepath"
	"sort"
	"strings"

	"github.com/fitan/genx/gen"
	"github.com/fitan/jennifer/jen"
	"github.com/samber/lo"
)

func New() *Plug {
	return &Plug{
		names: make(map[string]int, 0),
	}
}

type Plug struct {
	names map[string]int
}

func (p *Plug) getPkgName(name string) string {
	s, ok := p.names[name]
	if ok {
		p.names[name] = s + 1
		return fmt.Sprintf("%s%d", name, p.names[name])
	} else {
		p.names[name] = 0
		return name
	}
}

func (p *Plug) Name() string {
	return "@fx"
}

// FxStructPlug 处理 fx.In 和 fx.Out 结构体
type FxStructPlug struct {
	names map[string]int
}

func NewFxStructPlug() *FxStructPlug {
	return &FxStructPlug{
		names: make(map[string]int, 0),
	}
}

func (p *FxStructPlug) Name() string {
	return "@fx-struct"
}

func (p *FxStructPlug) getPkgName(name string) string {
	s, ok := p.names[name]
	if ok {
		p.names[name] = s + 1
		return fmt.Sprintf("%s%d", name, p.names[name])
	} else {
		p.names[name] = 0
		return name
	}
}

func (p *FxStructPlug) Gen(option gen.Option, structGoTypeMetas []gen.StructGoTypeMeta) (res []gen.GenResult, err error) {
	f := jen.NewFile(option.Pkg.Name)
	f.AddImport("go.uber.org/fx", "")

	hasGenerated := false

	for _, v := range structGoTypeMetas {
		// 检查是否有 @fx-struct 注解
		fxDoc := v.Doc.ByFuncName(p.Name())
		if fxDoc == nil {
			continue
		}

		// 解析参数
		var structType string
		if len(fxDoc.Args) > 0 {
			structType = strings.Trim(fxDoc.Args[0].Value, `"`)
		}

		slog.Info("processing fx struct",
			"struct", v.Name,
			"type", structType)

		switch structType {
		case "in":
			p.generateFxInStruct(f, v)
			hasGenerated = true
		case "out":
			p.generateFxOutStruct(f, v)
			hasGenerated = true
		default:
			slog.Error("unsupported fx struct type",
				"struct", v.Name,
				"type", structType,
				"supported_types", "in|out")
		}
	}

	if hasGenerated {
		res = append(res, gen.GenResult{
			FileName: filepath.Join(option.Dir, "fx_structs.go"),
			FileStr:  f.GoString(),
			Cover:    true,
		})
	}

	return
}

func (p *FxStructPlug) generateFxInStruct(f *jen.File, meta gen.StructGoTypeMeta) {
	// 生成 fx.In 结构体
	f.Type().Id(meta.Name + "In").StructFunc(func(g *jen.Group) {
		g.Id("fx.In")

		// 遍历原结构体的字段，生成对应的依赖注入字段
		for i := 0; i < meta.Obj.NumFields(); i++ {
			field := meta.Obj.Field(i)
			if !field.IsField() {
				continue
			}

			// 获取字段类型
			fieldType := field.Type().String()

			// 检查字段是否有特殊标签
			tag := meta.Obj.Tag(i)
			var fxTag string

			if strings.Contains(tag, `group:`) {
				// 提取group标签
				fxTag = tag
			} else if strings.Contains(tag, `name:`) {
				// 提取name标签
				fxTag = tag
			}

			if fxTag != "" {
				g.Id(field.Name()).Id(fieldType).Tag(map[string]string{"fx": fxTag})
			} else {
				g.Id(field.Name()).Id(fieldType)
			}
		}
	}).Line()
}

func (p *FxStructPlug) generateFxOutStruct(f *jen.File, meta gen.StructGoTypeMeta) {
	// 生成 fx.Out 结构体
	f.Type().Id(meta.Name + "Out").StructFunc(func(g *jen.Group) {
		g.Id("fx.Out")

		// 遍历原结构体的字段，生成对应的输出字段
		for i := 0; i < meta.Obj.NumFields(); i++ {
			field := meta.Obj.Field(i)
			if !field.IsField() {
				continue
			}

			// 获取字段类型
			fieldType := field.Type().String()

			// 检查字段是否有特殊标签
			tag := meta.Obj.Tag(i)
			var fxTag string

			if strings.Contains(tag, `group:`) {
				// 提取group标签
				fxTag = tag
			} else if strings.Contains(tag, `name:`) {
				// 提取name标签
				fxTag = tag
			}

			if fxTag != "" {
				g.Id(field.Name()).Id(fieldType).Tag(map[string]string{"fx": fxTag})
			} else {
				g.Id(field.Name()).Id(fieldType)
			}
		}
	}).Line()
}

func (p *Plug) Gen(req []gen.GlobalFuncGoTypeMeta) (res []gen.GenResult, err error) {
	slog.Info("fx plugin starting", "total_packages", len(req))

	appMetas := []gen.GlobalFuncGoTypeMeta{}
	provideMetas := []provideMeta{}
	invokeMetas := []invokeMeta{}
	decorateMetas := []decorateMeta{}

	// 扫描所有函数，按类型分类
	lo.ForEach(req, func(item gen.GlobalFuncGoTypeMeta, index int) {
		slog.Info("processing package", "package", item.Option.Pkg.PkgPath, "functions", len(item.Metas))
		appMeta := gen.GlobalFuncGoTypeMeta{
			Option: item.Option,
			Metas:  []gen.FuncGoTypeMeta{},
		}
		provideMeta := provideMeta{
			Option: item.Option,
			Metas:  []provideMetaItem{},
		}
		invokeMeta := invokeMeta{
			Option: item.Option,
			Metas:  []invokeMetaItem{},
		}
		decorateMeta := decorateMeta{
			Option: item.Option,
			Metas:  []decorateMetaItem{},
		}

		var appMetaOk, provideMetaOk, invokeMetaOk, decorateMetaOk bool

		lo.ForEach(item.Metas, func(meta gen.FuncGoTypeMeta, index int) {
			// 检查是否有 @fx 注解
			fxDoc := meta.Doc.ByFuncName(p.Name())
			if fxDoc == nil {
				return // 没有 @fx 注解，跳过
			}

			// 解析键值对参数
			var fxType, fxName, fxGroup string

			// 从参数中提取 type
			if typeValue, ok := meta.Doc.ByFuncNameAndArgName(p.Name(), "type"); ok {
				fxType = typeValue
			}

			// 从参数中提取 name (接口名)
			if nameValue, ok := meta.Doc.ByFuncNameAndArgName(p.Name(), "name"); ok {
				fxName = nameValue
			}

			// 从参数中提取 group
			if groupValue, ok := meta.Doc.ByFuncNameAndArgName(p.Name(), "group"); ok {
				fxGroup = groupValue
			}

			slog.Info("processing fx function",
				"function", meta.Name,
				"type", fxType,
				"name", fxName,
				"group", fxGroup)

			switch fxType {
			case "app":
				appMeta.Metas = append(appMeta.Metas, meta)
				appMetaOk = true
			case "provide":
				provideMeta.Metas = append(provideMeta.Metas, provideMetaItem{
					FxName:  fxName,
					FxGroup: fxGroup,
					Meta:    meta,
				})
				provideMetaOk = true
			case "invoke":
				invokeMeta.Metas = append(invokeMeta.Metas, invokeMetaItem{
					FxName: fxName,
					Meta:   meta,
				})
				invokeMetaOk = true
			case "decorate":
				decorateMeta.Metas = append(decorateMeta.Metas, decorateMetaItem{
					FxName: fxName,
					Meta:   meta,
				})
				decorateMetaOk = true
			default:
				if fxType != "" {
					slog.Error("unsupported fx function type",
						"function", meta.Name,
						"fx_type", fxType,
						"supported_types", "app|provide|invoke|decorate")
				}
			}
		})

		if appMetaOk {
			appMetas = append(appMetas, appMeta)
		}
		if provideMetaOk {
			provideMetas = append(provideMetas, provideMeta)
		}
		if invokeMetaOk {
			invokeMetas = append(invokeMetas, invokeMeta)
		}
		if decorateMetaOk {
			decorateMetas = append(decorateMetas, decorateMeta)
		}
	})

	// 按包路径排序，确保生成代码的一致性
	sort.Slice(provideMetas, func(i, j int) bool {
		return provideMetas[i].Option.Pkg.PkgPath > provideMetas[j].Option.Pkg.PkgPath
	})
	sort.Slice(invokeMetas, func(i, j int) bool {
		return invokeMetas[i].Option.Pkg.PkgPath > invokeMetas[j].Option.Pkg.PkgPath
	})
	sort.Slice(decorateMetas, func(i, j int) bool {
		return decorateMetas[i].Option.Pkg.PkgPath > decorateMetas[j].Option.Pkg.PkgPath
	})

	// 为每个标记为 app 的包生成 fx_app.go
	lo.ForEach(appMetas, func(item gen.GlobalFuncGoTypeMeta, index int) {
		f := jen.NewFile(item.Option.Pkg.Name)
		f.AddImport("go.uber.org/fx", "")

		// 生成 FxOptions 函数
		f.Func().Id("FxOptions").Params().Index().Id("fx.Option").BlockFunc(func(g *jen.Group) {
			g.Return(jen.Index().Id("fx.Option").ValuesFunc(func(optGroup *jen.Group) {
				// 添加所有 Provide 选项
				p.addProvideOptions(f, optGroup, provideMetas)
				// 添加所有 Invoke 选项
				p.addInvokeOptions(f, optGroup, invokeMetas)
				// 添加所有 Decorate 选项
				p.addDecorateOptions(f, optGroup, decorateMetas)
			}))
		}).Line()

		// 生成便捷的 NewApp 函数
		f.Func().Id("NewApp").Params(jen.Id("opts").Op("...").Id("fx.Option")).Op("*").Id("fx.App").BlockFunc(func(g *jen.Group) {
			g.Id("allOpts").Op(":=").Append(jen.Id("FxOptions").Call(), jen.Id("opts").Op("..."))
			g.Return(jen.Id("fx.New").Call(jen.Id("allOpts").Op("...")))
		}).Line()

		res = append(res, gen.GenResult{
			FileName: filepath.Join(item.Option.Dir, "fx_app.go"),
			FileStr:  f.GoString(),
			Cover:    true,
		})
	})

	return
}

// 数据结构定义
type provideMeta struct {
	Option gen.Option
	Metas  []provideMetaItem
}

type provideMetaItem struct {
	FxName  string
	FxGroup string
	Meta    gen.FuncGoTypeMeta
}

type invokeMeta struct {
	Option gen.Option
	Metas  []invokeMetaItem
}

type invokeMetaItem struct {
	FxName string
	Meta   gen.FuncGoTypeMeta
}

type decorateMeta struct {
	Option gen.Option
	Metas  []decorateMetaItem
}

type decorateMetaItem struct {
	FxName string
	Meta   gen.FuncGoTypeMeta
}

// 添加 Provide 选项的辅助方法
func (p *Plug) addProvideOptions(f *jen.File, optGroup *jen.Group, provideMetas []provideMeta) {
	lo.ForEach(provideMetas, func(v provideMeta, index int) {
		pkgPath := v.Option.Pkg.PkgPath
		pkgName := v.Option.Pkg.Name
		pkgName = p.getPkgName(pkgName)
		f.AddImport(pkgPath, pkgName)

		sort.Slice(v.Metas, func(i, j int) bool {
			return v.Metas[i].Meta.Name > v.Metas[j].Meta.Name
		})

		lo.ForEach(v.Metas, func(vv provideMetaItem, index int) {
			if vv.FxGroup != "" {
				// fx.Provide(fx.Annotate(constructor, fx.ResultTags(`group:"groupname"`)))
				optGroup.Id("fx.Provide").Call(
					jen.Id("fx.Annotate").Call(
						jen.Id(pkgName+"."+vv.Meta.Name),
						jen.Id("fx.ResultTags").Call(jen.Lit(fmt.Sprintf(`group:"%s"`, vv.FxGroup))),
					),
				)
			} else if vv.FxName != "" {
				// fx.Provide(fx.Annotate(constructor, fx.As(new(Interface))))
				optGroup.Id("fx.Provide").Call(
					jen.Id("fx.Annotate").Call(
						jen.Id(pkgName+"."+vv.Meta.Name),
						jen.Id("fx.As").Call(jen.New(jen.Id(vv.FxName))),
					),
				)
			} else {
				// fx.Provide(constructor)
				optGroup.Id("fx.Provide").Call(jen.Id(pkgName + "." + vv.Meta.Name))
			}
		})
	})
}

// 添加 Invoke 选项的辅助方法
func (p *Plug) addInvokeOptions(f *jen.File, optGroup *jen.Group, invokeMetas []invokeMeta) {
	lo.ForEach(invokeMetas, func(v invokeMeta, index int) {
		pkgPath := v.Option.Pkg.PkgPath
		pkgName := v.Option.Pkg.Name
		pkgName = p.getPkgName(pkgName)
		f.AddImport(pkgPath, pkgName)

		sort.Slice(v.Metas, func(i, j int) bool {
			return v.Metas[i].Meta.Name > v.Metas[j].Meta.Name
		})

		lo.ForEach(v.Metas, func(vv invokeMetaItem, index int) {
			optGroup.Id("fx.Invoke").Call(jen.Id(pkgName + "." + vv.Meta.Name))
		})
	})
}

// 添加 Decorate 选项的辅助方法
func (p *Plug) addDecorateOptions(f *jen.File, optGroup *jen.Group, decorateMetas []decorateMeta) {
	lo.ForEach(decorateMetas, func(v decorateMeta, index int) {
		pkgPath := v.Option.Pkg.PkgPath
		pkgName := v.Option.Pkg.Name
		pkgName = p.getPkgName(pkgName)
		f.AddImport(pkgPath, pkgName)

		sort.Slice(v.Metas, func(i, j int) bool {
			return v.Metas[i].Meta.Name > v.Metas[j].Meta.Name
		})

		lo.ForEach(v.Metas, func(vv decorateMetaItem, index int) {
			if vv.FxName != "" {
				// fx.Decorate(fx.Annotate(decorator, fx.As(new(Interface))))
				optGroup.Id("fx.Decorate").Call(
					jen.Id("fx.Annotate").Call(
						jen.Id(pkgName+"."+vv.Meta.Name),
						jen.Id("fx.As").Call(jen.New(jen.Id(vv.FxName))),
					),
				)
			} else {
				// fx.Decorate(decorator)
				optGroup.Id("fx.Decorate").Call(jen.Id(pkgName + "." + vv.Meta.Name))
			}
		})
	})
}
