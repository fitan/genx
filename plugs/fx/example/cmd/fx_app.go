// Code generated . DO NOT EDIT.
package main

import "go.uber.org/fx"

func FxOptions() []fx.Option {
	return []fx.Option{}
}

func NewApp(opts ...fx.Option) *fx.App {
	allOpts := append(FxOptions(), opts...)
	return fx.New(allOpts...)
}
