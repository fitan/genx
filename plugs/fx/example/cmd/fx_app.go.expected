// 这是预期生成的代码示例
package main

import (
	"go.uber.org/fx"
	app "github.com/fitan/genx/plugs/fx/example/app"
	infrastructure "github.com/fitan/genx/plugs/fx/example/infrastructure"
	middleware "github.com/fitan/genx/plugs/fx/example/middleware"
	services "github.com/fitan/genx/plugs/fx/example/services"
)

// FxOptions 返回所有自动收集的 fx.Option
func FxOptions() []fx.Option {
	return []fx.Option{
		// Provide 选项 (按包路径和函数名排序)
		fx.Provide(infrastructure.NewDatabase),
		fx.Provide(infrastructure.NewLogger),
		
		fx.Provide(fx.Annotate(
			services.NewUserService,
			fx.As(new(services.UserService)),
		)),
		
		fx.Provide(fx.Annotate(
			services.NewUserHandler,
			fx.ResultTags(`group:"handlers"`),
		)),
		
		// Decorate 选项
		fx.Decorate(fx.Annotate(
			middleware.NewLoggingUserService,
			fx.As(new(services.UserService)),
		)),
		
		// Invoke 选项
		fx.Invoke(app.InitDatabase),
		fx.Invoke(app.StartApplication),
	}
}

// NewApp 创建配置好的 fx.App
func NewApp(opts ...fx.Option) *fx.App {
	allOpts := append(FxOptions(), opts...)
	return fx.New(allOpts...)
}
